File: leetcode/UniquePathsII/uniquePathsWithObstacles.java
public class Solution {
    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        // Start typing your Java solution below
        // DO NOT write main() function
        int m = obstacleGrid.length, n = obstacleGrid[0].length, sols[] = new int[n];
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                sols[j] = obstacleGrid[i][j] == 1 ? 0 : sols[j] + (j == 0 ? (i == 0 ? 1 : 0) : sols[j - 1]);
            }
        }
        return sols[n - 1];
    }
}

File: leetcode/UniquePathsII/uniquePathsWithObstacles.cpp
class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
        int m = obstacleGrid.size(), n = m ? obstacleGrid[0].size() : 0;
        vector<int> ans(n, 0);
        ans[0] = 1;
        for (auto row : obstacleGrid) {
            for (int i = 0; i < n; i++) {
                if (!row[i]) ans[i] += i == 0 ? 0 : ans[i - 1];
                else ans[i] = 0;
            }
        }
        return ans[n - 1];
    }
};
